api/src/main/java/org/apache/iceberg/UpdatePartitionSpec.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import org.apache.iceberg.exceptions.CommitFailedException;
+
+/**
+ * API for partition spec evolution.
+ * <p>
+ * When committing, these changes will be applied to the current table metadata. Commit conflicts
+ * will not be resolved and will result in a {@link CommitFailedException}.
+ */
+public interface UpdatePartitionSpec extends PendingUpdate<PartitionSpec> {
```

Comment:
Can we remove the new interface and methods from this PR?

I don't think this is needed for this PR, and I'd like to minimize the number of changes. In addition, I don't think we want to move to a model where users create a new spec and apply it to a table. I think we instead want to evolve the partition spec of a table. So this API will probably be different when we release that feature.

Reply:
I agree that it is better to move the update to a new PR, which addresses #281. 
The main reason I put it here is to have additional unit tests to make sure it works as expected.

Additionally, I think we may support the table partition spec evolution in two ways
```
table.updatePartitionSpec()
  .update(spec)
  .commit();
```
```
table.updatePartitionSpec().newSpec(schema)
  .identity(...)
  .bucket(...)
  ...
  .commit();
```
The first approach may be used if clients want to define a spec and manage it by their codes, e.g. use defined spec object in multiple places.

Reply:
For tests, you can use `TableMetadata` and `commit` directly:

```java
TableOperations ops = table.operations();
TableMetadata base = ops.current()
TableMetadata updated = base.updatePartitionSpec(newSpec);
ops.commit(base, updated);
```

Reply:
👌